### PR TITLE
refactor: Extract out common logic from Tensorboard BatchedRequestSender into a base class.

### PR DESCRIPTION
**Pull request by yfang1**
Exacted out common logic from the 3 BatchedRequestSenders into a base class. 

This will facilitate optimizations down the road.

As nice side effects this brings up the uploader's test coverage due to fewer lines of code, and allows blob references to be batched instead of sent one at a time.

unit test passed. Tried uploading all 3 types of data. Working well.

### CODE REVIEW

File: google/cloud/aiplatform/tensorboard/uploader.py [hunk 1]
```diff
@@ -690,7 +693,7 @@ def __init__(
         max_request_size: int,
         tracker: upload_tracker.UploadTracker,
     ):
-        """Constructor for _ScalarBatchedRequestSender.
+        """Constructor for _NumericBatchedRequestSender.
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/tensorboard/uploader.py:696]: Is the name in this comment supposed to align with the name of class?

File: google/cloud/aiplatform/tensorboard/uploader.py [hunk 2]
```diff
@@ -831,43 +834,79 @@ def _create_point(
...
         value: tf.compat.v1.Summary.Value,
+        metadata: tf.compat.v1.SummaryMetadata,
     ):
         """Adds a scalar point to the given tag, if there's space.
 
         Args:
           time_series_proto: TimeSeriesData proto to which to add a point.
           event: Enclosing `Event` proto with the step and wall time data.
-          value: Scalar `Summary.Value` proto with the actual scalar data.
+          value: `Summary.Value` proto.
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/tensorboard/uploader.py:844]: `metadata` arg missing

File: google/cloud/aiplatform/tensorboard/uploader.py [hunk 3]
```diff
@@ -891,153 +929,95 @@ def __init__(
...
-        self._tensor_bytes_skipped = 0
 
-    def add_event(
-        self,
-        event: tf.compat.v1.Event,
-        value: tf.compat.v1.Summary.Value,
-        metadata: tf.compat.v1.SummaryMetadata,
-    ):
-        """Attempts to add the given event to the current request.
+    def _get_tracker(self):
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/tensorboard/uploader.py:936]: Return type hint.

File: google/cloud/aiplatform/tensorboard/uploader.py [hunk 4]
```diff
@@ -891,153 +929,95 @@ def __init__(
...
-            # because we just flushed.
-            try:
-                self._add_event_internal(event, value, metadata)
-            except _OutOfSpaceError:
-                raise RuntimeError("add_event failed despite flush")
+    def _get_value_type(self):
+        return tensorboard_time_series.TensorboardTimeSeries.ValueType.SCALAR
 
-    def _add_event_internal(
+    def _create_data_point(
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/tensorboard/uploader.py:942]: Return type hint.

File: google/cloud/aiplatform/tensorboard/uploader.py [hunk 5]
```diff
@@ -891,153 +929,95 @@ def __init__(
...
-                tag_name,
-                lambda: tensorboard_time_series.TensorboardTimeSeries(
-                    display_name=tag_name,
-                    value_type=tensorboard_time_series.TensorboardTimeSeries.ValueType.TENSOR,
-                    plugin_name=metadata.plugin_data.plugin_name,
-                    plugin_data=metadata.plugin_data.content,
-                ),
-            ).name.split("/")[-1],
-            value_type=tensorboard_time_series.TensorboardTimeSeries.ValueType.TENSOR,
+    def _get_tracker(self):
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/tensorboard/uploader.py:1003]: Return type hint.

File: google/cloud/aiplatform/tensorboard/uploader.py [hunk 6]
```diff
@@ -891,153 +929,95 @@ def __init__(
...
 
-        self._request.time_series_data.extend([time_series_data_proto])
-        self._byte_budget_manager.add_time_series(time_series_data_proto)
-        self._tag_to_time_series_data[tag_name] = time_series_data_proto
-        return time_series_data_proto
+    def _get_value_type(self):
+        return tensorboard_time_series.TensorboardTimeSeries.ValueType.TENSOR
 
-    def _create_point(
+    def _create_data_point(
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/tensorboard/uploader.py:1014]: Return type hint.

File: google/cloud/aiplatform/tensorboard/uploader.py [hunk 7]
```diff
@@ -1198,73 +1171,49 @@ def __init__(
...
+        self._blob_sizes = 0
 
-    def add_event(
+    def _get_tracker(self):
+        return self._tracker.blob_tracker(0)
+
+    def _get_value_type(self):
+        return tensorboard_time_series.TensorboardTimeSeries.ValueType.BLOB_SEQUENCE
+
+    def _create_data_point(
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/tensorboard/uploader.py:1191]: Return type hint.

File: google/cloud/aiplatform/tensorboard/uploader.py [hunk 8]
```diff
@@ -831,43 +834,79 @@ def _create_point(
...
+    def _get_tracker(self):
+        """
+        :return: tracker function from upload_tracker.UploadTracker
+        """
+        pass
 
-class _TensorBatchedRequestSender(object):
-    """Helper class for building WriteTensor() requests that fit under a size limit.
+    @abc.abstractmethod
+    def _get_value_type(
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/tensorboard/uploader.py:870]: Based on the usage of this method, it seems like it should be an class property. ie:

```Python
class _BaseBatchRequestSender:
    ...
    @property
    @classmethod
    @abc.abstractmethod
    def _value_type(cls) -> tensorboard_time_series.TensorboardTimeSeries.ValueType:
        pass


class _ScalarBatchedRequestSender(_BaseBatchedRequestSender):
    
    _value_type = tensorboard_time_series.TensorboardTimeSeries.ValueType.SCALAR
    ....

_ScalarBatchedRequestSender._value_type
```

**Review by sasha-gitg — APPROVED** [review]
LGTM. Left a few comments. Thanks!

File: google/cloud/aiplatform/tensorboard/uploader.py (hunk 5 above)
Comment (yfang1) [reply google/cloud/aiplatform/tensorboard/uploader.py]: added

File: google/cloud/aiplatform/tensorboard/uploader.py (hunk 4 above)
Comment (yfang1) [reply google/cloud/aiplatform/tensorboard/uploader.py:950]: added

File: google/cloud/aiplatform/tensorboard/uploader.py (hunk 3 above)
Comment (yfang1) [reply google/cloud/aiplatform/tensorboard/uploader.py]: added

File: google/cloud/aiplatform/tensorboard/uploader.py (hunk 8 above)
Comment (yfang1) [reply google/cloud/aiplatform/tensorboard/uploader.py]: awesome. Changed.

File: google/cloud/aiplatform/tensorboard/uploader.py (hunk 2 above)
Comment (yfang1) [reply google/cloud/aiplatform/tensorboard/uploader.py:852]: fixed

File: google/cloud/aiplatform/tensorboard/uploader.py (hunk 1 above)
Comment (yfang1) [reply google/cloud/aiplatform/tensorboard/uploader.py]: I supposed so. Forgot to update the comment when renaming for the base class. Fixed.

**Review by vinnysenthil — APPROVED** [review]

**Pull request merged**